storage/innobase/buf/buf0buf.cc

```diff
@@ -1878,6 +2092,71 @@ inline void buf_pool_t::resize()
 	return;
 }
 
+inline void buf_pool_t::garbage_collect()
+{
+  mysql_mutex_lock(&mutex);
+  lru_hp.set(nullptr);
+  size_t freed= 0;
+
+  for (buf_page_t *bpage= UT_LIST_GET_LAST(LRU), *prev; bpage; bpage= prev)
+  {
+    prev= UT_LIST_GET_PREV(LRU, bpage);
+    auto state= bpage->state();
+    ut_ad(state >= buf_page_t::FREED);
+    ut_ad(bpage->in_LRU_list);
+
+    /* We try to free any pages that can be freed without writing out
+    anything. */
+    switch (bpage->oldest_modification()) {
+    case 0:
+    try_to_evict:
+      if (buf_LRU_free_page(bpage, true))
+        freed++;
+      continue;
```

Comment:
Sorry, I had forgotten that this may temporarily release `buf_pool.mutex`. Please revise as follows:
```diff
@@ -2095,12 +2095,22 @@ inline void buf_pool_t::resize()
 inline void buf_pool_t::garbage_collect()
 {
   mysql_mutex_lock(&mutex);
-  lru_hp.set(nullptr);
   size_t freed= 0;
 
+#ifdef BTR_CUR_HASH_ADAPT
+  /* buf_LRU_free_page() will temporarily release and reacquire
+  buf_pool.mutex for invoking btr_search_drop_page_hash_index(). Thus,
+  we must protect ourselves with the hazard pointer. */
+rescan:
+#else
+  lru_hp.set(nullptr);
+#endif
   for (buf_page_t *bpage= UT_LIST_GET_LAST(LRU), *prev; bpage; bpage= prev)
   {
     prev= UT_LIST_GET_PREV(LRU, bpage);
+#ifdef BTR_CUR_HASH_ADAPT
+    buf_pool.lru_hp.set(prev);
+#endif
     auto state= bpage->state();
     ut_ad(state >= buf_page_t::FREED);
     ut_ad(bpage->in_LRU_list);
@@ -2111,7 +2121,16 @@ inline void buf_pool_t::garbage_collect()
     case 0:
     try_to_evict:
       if (buf_LRU_free_page(bpage, true))
+      {
+      evicted:
         freed++;
+#ifdef BTR_CUR_HASH_ADAPT
+        bpage= prev;
+        prev= buf_pool.lru_hp.get();
+        if (!prev && bpage)
+          goto rescan;
+#endif
+      }
       continue;
     case 1:
       break;
@@ -2136,7 +2155,10 @@ inline void buf_pool_t::garbage_collect()
       default:
         if (bpage->state() < buf_page_t::UNFIXED &&
             oldest_modification <= log_sys.get_flushed_lsn())
+        {
           release_freed_page(bpage);
+          goto evicted;
+        }
         else
           bpage->lock.u_unlock(true);
       }
```
This will also make us count marked-as-freed pages that were removed both from `flush_list`  and `LRU`.